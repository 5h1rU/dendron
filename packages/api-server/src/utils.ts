import { DEngineV2 } from "@dendronhq/common-all/src";
import { DendronEngineV2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { MemoryStore } from "./store/memoryStore";

export function getWSKey(uri: string) {
  return _.trimEnd(uri, "/");
}

export async function putWS({
  ws,
  engine,
}: {
  ws: string;
  engine: DendronEngineV2;
}) {
  MemoryStore.instance().put(`ws:${getWSKey(ws)}`, engine);
}

export async function getWS({ ws }: { ws: string }) {
  const engine = await MemoryStore.instance().get<DEngineV2>(
    `ws:${getWSKey(ws)}`
  );
  if (!engine) {
    throw `No Engine: ${ws}`;
  }
  return engine;
}
