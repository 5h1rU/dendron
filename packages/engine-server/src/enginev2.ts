import {
  ConfigWriteOpts,
  DendronConfig,
  DendronError,
  DEngineClientV2,
  DEngineDeleteSchemaRespV2,
  DEngineInitRespV2,
  DEngineMode,
  DEngineV2,
  DLink,
  DNodeTypeV2,
  DStoreV2,
  DVault,
  EngineDeleteOptsV2,
  EngineUpdateNodesOptsV2,
  EngineWriteOptsV2,
  GetNoteOptsV2,
  GetNotePayloadV2,
  NoteChangeEntry,
  NotePropsDictV2,
  NotePropsV2,
  NoteUtilsV2,
  QueryNotesOpts,
  RenameNoteOptsV2,
  RenameNotePayload,
  RespV2,
  SchemaModuleDictV2,
  SchemaModulePropsV2,
  SchemaQueryResp,
  WorkspaceOpts,
  WriteNoteResp,
} from "@dendronhq/common-all";
import {
  createLogger,
  DLogger,
  readYAML,
  VaultUtils,
  writeYAML,
} from "@dendronhq/common-server";
import _ from "lodash";
import { DConfig } from "./config";
import { FileStorageV2 } from "./drivers/file/storev2";
import { FuseEngine } from "./fuseEngine";

type DendronEngineOptsV2 = {
  wsRoot: string;
  vaultsv3: DVault[];
  forceNew?: boolean;
  store?: any;
  mode?: DEngineMode;
  logger?: DLogger;
};
type DendronEnginePropsV2 = Required<DendronEngineOptsV2>;

export class DendronEngineV2 implements DEngineV2 {
  public wsRoot: string;
  public vaults: string[];
  public store: DStoreV2;
  protected props: DendronEnginePropsV2;
  public logger: DLogger;
  public fuseEngine: FuseEngine;
  public links: DLink[];
  public vaultsv3: DVault[];
  public configRoot: string;

  static _instance: DendronEngineV2 | undefined;

  constructor(props: DendronEnginePropsV2) {
    this.wsRoot = props.wsRoot;
    this.configRoot = props.wsRoot;
    this.store = props.store;
    this.logger = props.logger;
    this.props = props;
    this.fuseEngine = new FuseEngine({});
    this.links = [];
    this.vaultsv3 = props.vaultsv3;
    this.vaults = props.vaultsv3.map((ent) => ent.fsPath);
  }

  static createV3({
    vaults,
    wsRoot,
    logger,
  }: WorkspaceOpts & { logger?: DLogger }) {
    const LOGGER = logger || createLogger();
    return new DendronEngineV2({
      wsRoot,
      vaultsv3: vaults,
      forceNew: true,
      store: new FileStorageV2({
        wsRoot,
        vaultsv3: vaults,
        logger: LOGGER,
      }),
      mode: "fuzzy",
      logger: LOGGER,
    });
  }

  static instance({ vaults, wsRoot }: { vaults: DVault[]; wsRoot: string }) {
    if (!DendronEngineV2._instance) {
      DendronEngineV2._instance = DendronEngineV2.createV3({ vaults, wsRoot });
    }
    return DendronEngineV2._instance;
  }

  get notes(): NotePropsDictV2 {
    return this.store.notes;
  }
  get schemas(): SchemaModuleDictV2 {
    return this.store.schemas;
  }

  /**
   * Does not throw error but returns it
   */
  async init(): Promise<DEngineInitRespV2> {
    try {
      const { data, error } = await this.store.init();
      const { notes, schemas } = data;
      this.updateIndex("note");
      this.updateIndex("schema");
      return {
        error,
        data: { notes, schemas },
      };
    } catch (error) {
      const { message, stack, msg, status, friendly } = error;
      let payload = { message, stack };
      return {
        error: new DendronError({ payload, msg, status, friendly }),
        data: {
          notes: {},
          schemas: {},
        },
      };
    }
  }

  async deleteNote(
    id: string,
    opts?: EngineDeleteOptsV2
  ): ReturnType<DEngineClientV2["deleteNote"]> {
    try {
      const note = this.notes[id];
      const changed = await this.store.deleteNote(id, opts);
      const noteChangeEntry = _.find(
        changed,
        (ent) => ent.note.id === id
      ) as NoteChangeEntry;
      if (noteChangeEntry.status === "delete") {
        await this.fuseEngine.removeNoteFromIndex(note);
      }
      return {
        data: changed,
        error: null,
      };
    } catch (err) {
      return {
        data: [],
        error: err,
      };
    }
  }

  async deleteSchema(
    id: string,
    opts?: EngineDeleteOptsV2
  ): Promise<DEngineDeleteSchemaRespV2> {
    try {
      const data = (await this.store.deleteSchema(
        id,
        opts
      )) as DEngineDeleteSchemaRespV2;
      // deleted schema might affect notes
      await this.updateIndex("note");
      await this.updateIndex("schema");
      return data;
      // FIXM:E not performant
      // const smod = this.schemas[id];
      // await this.fuseEngine.removeSchemaFromIndex(smod);
      // return {
      //   data: undefined,
      //   error: null,
      // };
    } catch (err) {
      return {
        data: {
          notes: {},
          schemas: {},
        },
        error: err,
      };
    }
  }

  async getNoteByPath({
    npath,
    createIfNew,
    vault,
  }: GetNoteOptsV2): Promise<RespV2<GetNotePayloadV2>> {
    const ctx = "getNoteByPath";
    this.logger.debug({ ctx, npath, createIfNew, msg: "enter" });
    const maybeNote = NoteUtilsV2.getNoteByFname(npath, this.notes, {
      vault: vault,
    });
    this.logger.debug({ ctx, maybeNote, msg: "post-query" });
    let noteNew: NotePropsV2 | undefined = maybeNote;
    let changed: NoteChangeEntry[] = [];
    let error = null;
    if ((!maybeNote || maybeNote.stub) && createIfNew) {
      this.logger.debug({ ctx, maybeNote, msg: "create-new" });
      if (maybeNote?.stub) {
        noteNew = maybeNote;
        delete noteNew.stub;
      } else {
        noteNew = NoteUtilsV2.createWithSchema({
          noteOpts: { fname: npath, vault },
          engine: this,
        });
      }
      changed = (await this.writeNote(noteNew, { newNode: true })).data;
    }
    if (!createIfNew && !maybeNote) {
      error = new DendronError({ status: "no_note_found" });
    }
    return {
      data: { note: noteNew, changed },
      error,
    };
  }

  async getConfig() {
    const cpath = DConfig.configPath(this.configRoot);
    const config = readYAML(cpath) as DendronConfig;
    return {
      error: null,
      data: config,
    };
  }

  async getSchema(id: string): Promise<RespV2<SchemaModulePropsV2>> {
    const ctx = "getSchema";
    const data = this.schemas[id];
    this.logger.info({ ctx, msg: "exit" });
    return {
      data,
      error: null,
    };
  }

  queryNotesSync({
    qs,
  }: {
    qs: string;
  }): ReturnType<DEngineClientV2["queryNotesSync"]> {
    const items = this.fuseEngine.queryNote({ qs });
    return {
      error: null,
      data: items.map((ent) => this.notes[ent.id]),
    };
  }

  async querySchema(queryString: string): Promise<SchemaQueryResp> {
    const ctx = "querySchema";

    let items: SchemaModulePropsV2[] = [];
    const results = await this.fuseEngine.querySchema({ qs: queryString });
    items = results.map((ent) => this.schemas[ent.id]);
    // if (queryString === "") {
    //   items = [this.schemas.root];
    // } else if (queryString === "*") {
    //   items = _.values(this.schemas);
    // } else {
    //   const results = this.schemaIndex.search(queryString);
    //   items = _.map(results, (resp) => this.schemas[resp.item.id]);
    // }
    this.logger.info({ ctx, msg: "exit" });
    return {
      error: null,
      data: items,
    };
  }

  async queryNotes(
    opts: QueryNotesOpts
  ): ReturnType<DEngineClientV2["queryNotes"]> {
    const ctx = "Engine:queryNotes";
    const { qs, vault, createIfNew } = opts;
    let items = await this.fuseEngine.queryNote({ qs });
    let item = this.notes[items[0].id];
    if (createIfNew) {
      let noteNew: NotePropsV2;
      if (item?.fname === qs && item?.stub) {
        noteNew = item;
        noteNew.stub = false;
      } else {
        if (_.isUndefined(vault)) {
          return {
            error: new DendronError({ msg: "no vault specified" }),
            data: null as any,
          };
        }
        noteNew = NoteUtilsV2.create({ fname: qs, vault });
      }
      await this.writeNote(noteNew, { newNode: true });
    }
    this.logger.info({ ctx, msg: "exit" });
    let notes = items.map((ent) => this.notes[ent.id]);
    if (!_.isUndefined(vault)) {
      notes = notes.filter((ent) => VaultUtils.isEqual(vault, ent.vault));
    }
    return {
      error: null,
      data: notes,
    };
  }

  // async query(
  //   queryString: string,
  //   mode: DNodeTypeV2,
  //   opts?: QueryOptsV2
  // ): Promise<EngineQueryNoteResp> {
  //   const ctx = "Engine:query";
  //   const cleanOpts = _.defaults(opts || {}, {
  //     fullNode: false,
  //     createIfNew: false,
  //     initialQuery: false,
  //     stub: false,
  //   });
  //   this.logger.info({ ctx, msg: "enter" });
  //   let items: DNodePropsV2[] = [];

  //   // ~~~ schema query
  //   if (mode === "schema") {
  //     throw Error("engine.query for schema is not supported");
  //   } else {
  //     // ~~~ note query
  //     items = await this.fuseEngine.queryNote({ qs: queryString });
  //     // if (queryString === "") {
  //     //   items = [this.notes.root];
  //     // } else if (queryString === "*") {
  //     //   items = _.values(this.notes);
  //     // } else {
  //     //   const results = this.notesIndex.search(queryString);
  //     //   items = _.map(results, (resp) => resp.item);
  //     // }
  //     if (cleanOpts.createIfNew) {
  //       let noteNew: NotePropsV2;
  //       if (items[0]?.fname === queryString && items[0]?.stub) {
  //         noteNew = items[0];
  //         noteNew.stub = false;
  //       } else {
  //         noteNew = NoteUtilsV2.create({ fname: queryString });
  //       }
  //       await this.writeNote(noteNew, { newNode: true });
  //     }
  //     if (cleanOpts.fullNode) {
  //       throw Error("not implemented");
  //     }
  //   }

  //   // ~~~ exit
  //   this.logger.info({ ctx, msg: "exit" });
  //   return {
  //     error: null,
  //     data: items,
  //   };
  // }

  async sync() {
    throw Error("sync not implemented");
    return {} as any;
  }

  async renameNote(opts: RenameNoteOptsV2): Promise<RespV2<RenameNotePayload>> {
    const resp = await this.store.renameNote(opts);
    return {
      error: null,
      data: resp,
    };
  }

  async updateNote(
    note: NotePropsV2,
    opts?: EngineUpdateNodesOptsV2
  ): Promise<void> {
    const out = this.store.updateNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async updateIndex(mode: DNodeTypeV2) {
    if (mode === "schema") {
      this.fuseEngine.updateSchemaIndex(this.schemas);
    } else {
      this.fuseEngine.updateNotesIndex(this.notes);
    }
  }

  async updateSchema(schemaModule: SchemaModulePropsV2) {
    const out = await this.store.updateSchema(schemaModule);
    await this.updateIndex("schema");
    return out;
  }

  async writeConfig(
    opts: ConfigWriteOpts
  ): ReturnType<DEngineV2["writeConfig"]> {
    const { configRoot } = this;
    const cpath = DConfig.configPath(configRoot);
    writeYAML(cpath, opts.config);
    return {
      error: null,
    };
  }

  async writeNote(
    note: NotePropsV2,
    opts?: EngineWriteOptsV2
  ): Promise<WriteNoteResp> {
    const out = await this.store.writeNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async writeSchema(schema: SchemaModulePropsV2) {
    return this.store.writeSchema(schema);
  }
}
