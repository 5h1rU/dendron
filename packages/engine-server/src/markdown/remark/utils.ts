import {
  CONSTANTS,
  DendronError,
  getSlugger,
  NoteChangeEntry,
  NoteProps,
  NoteUtils,
} from "@dendronhq/common-all";
import { normalizev2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { Heading, Root } from "mdast";
import { Processor } from "unified";
import { Node } from "unist";
import { selectAll } from "unist-util-select";
import { VFile } from "vfile";
import { WikiLinkProps } from "../../topics/markdown";
import {
  DendronASTRoot,
  DendronASTTypes,
  NoteRefNoteV4_LEGACY,
} from "../types";

export const ALIAS_DIVIDER = "|";

export function addError(proc: Processor, err: DendronError) {
  const errors = proc.data("errors") as DendronError[];
  errors.push(err);
  proc().data("errors", errors);
}

export function getNoteOrError(
  notes: NoteProps[],
  hint: any
): { error: DendronError | undefined; note: undefined | NoteProps } {
  let error: DendronError | undefined;
  let note: NoteProps | undefined;
  if (_.isUndefined(notes)) {
    error = new DendronError({ msg: `no note found. ${hint}` });
    return { error, note };
  }
  if (notes.length > 1) {
    error = new DendronError({ msg: `multiple notes found for link: ${hint}` });
    return { error, note };
  }
  if (notes.length < 1) {
    error = new DendronError({
      msg: `no notes found for link: ${JSON.stringify(hint)}`,
    });
    return { error, note };
  }
  note = notes[0];
  return { error, note };
}

export class LinkUtils {
  static isAlias(link: string) {
    return link.indexOf("|") !== -1;
  }

  static hasFilter(link: string) {
    return link.indexOf(">") !== -1;
  }

  static parseAliasLink(link: string) {
    const [alias, value] = link.split("|").map(_.trim);
    return { alias, value: NoteUtils.normalizeFname(value) };
  }

  static parseDendronURI(linkString: string) {
    if (linkString.startsWith(CONSTANTS.DENDRON_DELIMETER)) {
      const [vaultName, link] = linkString
        .split(CONSTANTS.DENDRON_DELIMETER)[1]
        .split("/");
      return {
        vaultName,
        link,
      };
    }
    return {
      link: linkString,
    };
  }

  static parseLinkV2(linkString: string) {
    const LINK_NAME = "[^#\\|>]+";
    const re = new RegExp(
      "" +
        // alias?
        `(` +
        `(?<alias>${LINK_NAME}(?=\\|))` +
        "\\|" +
        ")?" +
        // name
        `(?<value>${LINK_NAME})` +
        // anchor?
        `(#(?<anchor>${LINK_NAME}))?` +
        // filters?
        `(>(?<filtersRaw>.*))?`,
      "i"
    );
    const out = linkString.match(re);
    if (out) {
      let { alias, value, anchor } = out.groups as any;
      let vaultName: string | undefined;
      ({ vaultName, link: value } = this.parseDendronURI(value));
      if (!alias) {
        alias = value;
      }
      alias = _.trim(alias);
      value = _.trim(value);
      return { alias, value, anchorHeader: anchor, vaultName };
    } else {
      return null;
    }
  }

  static parseLink(linkMatch: string) {
    linkMatch = NoteUtils.normalizeFname(linkMatch);
    let out: WikiLinkProps = {
      value: linkMatch,
      alias: linkMatch,
    };
    if (LinkUtils.isAlias(linkMatch)) {
      out = LinkUtils.parseAliasLink(linkMatch);
    }
    if (out.value.indexOf("#") !== -1) {
      const [value, anchorHeader] = out.value.split("#").map(_.trim);
      out.value = value;
      out.anchorHeader = anchorHeader;
      // if we didn't have an alias, links with a # anchor shouldn't have # portion be in the title
      if (!LinkUtils.isAlias(linkMatch)) {
        out.alias = value;
      }
    }
    return out;
  }
}

export class RemarkUtils {
  static oldNoteRef2NewNoteRef(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as DendronASTRoot;
        //@ts-ignore
        let notesRefLegacy: NoteRefNoteV4_LEGACY[] = selectAll("refLink", root);
        const slugger = getSlugger();
        notesRefLegacy.map((noteRefLegacy) => {
          // @ts-ignore;
          noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
          const { anchorStart, anchorEnd } = noteRefLegacy.data.link.data;
          if (anchorStart) {
            noteRefLegacy.data.link.data.anchorStart = normalizev2(
              anchorStart,
              slugger
            );
          }
          if (anchorEnd) {
            noteRefLegacy.data.link.data.anchorEnd = normalizev2(
              anchorEnd,
              slugger
            );
          }
        });
        changes.push({
          note,
          status: "update",
        });
        // if (out2.length > 0) {
        //   // const noteRefLegacy = root.children.splice(idx, 1)[0] as unknown as  NoteRefNoteV4_LEGACY;
        //   const noteRefLegacy = out2[0];
        //   // root.children[idx] as unknown as NoteRefNoteV4_LEGACY;
        //   // const noteRefNew: NoteRefNoteV4= {
        //   //   type: DendronASTTypes.REF_LINK_V2,
        //   //   data: {
        //   //     link: noteRefLegacy.data.link,
        //   //   },
        //   //   value: noteRefLegacy.value
        //   // }
        //   // @ts-ignore;
        //   noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
        //   // // @ts-ignore
        //   // root.children.splice(idx, 1, noteRefNew);
        //   changes.push({
        //     note,
        //     status: "update",
        //   });
        // }
      };
    };
  }

  static h1ToTitle(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children.splice(idx, 1)[0] as Heading;
          if (head.children.length === 1 && head.children[0].type === "text") {
            note.title = head.children[0].value;
          }
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }

  static h1ToH2(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children[idx] as Heading;
          head.depth = 2;
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }
}
