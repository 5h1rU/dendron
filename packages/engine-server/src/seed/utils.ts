import { DendronError, ERROR_STATUS, SeedConfig } from "@dendronhq/common-all";
import { simpleGit } from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { WorkspaceService } from "../workspace";
import { DEFAULT_SEED_PUBLISHER } from "./constants";

export class SeedUtils {
  static genDefaultConfig(opts: {
    seed?: SeedConfig;
    name: string;
  }): SeedConfig {
    return _.defaults(opts.seed || {}, {
      name: opts.name,
      publisher: DEFAULT_SEED_PUBLISHER,
      description: "a seed waiting to sprout",
      license: "CC BY 4.0",
      repository: {
        type: "git" as const,
        url: "",
      },
      root: "vault",
    });
  }

  static async clone({
    wsRoot,
    config,
  }: {
    config: SeedConfig;
    wsRoot: string;
  }) {
    const id = this.getSeedId(config);
    const spath = this.seed2Path({ wsRoot, id });
    fs.ensureDirSync(path.dirname(spath));
    const git = simpleGit({ baseDir: wsRoot });
    await git.clone(config.repository.url, spath);
    return spath;
  }

  static getSeedId({ publisher, name }: { publisher: string; name: string }) {
    return `${publisher}.${name}`;
  }

  static seed2Path({ wsRoot, id }: { wsRoot: string; id: string }) {
    return path.join(wsRoot, "seeds", id);
  }

  static validateWorkspaceSeedConversion({ wsRoot }: { wsRoot: string }) {
    const ws = new WorkspaceService({ wsRoot });
    if (ws.config.vaults.length !== 1) {
      return {
        error: DendronError.createFromStatus({
          status: ERROR_STATUS.INVALID_STATE,
          message: "workspace must have exactly one vault",
        }),
      };
    }
    if (!_.isEmpty(ws.config.workspaces)) {
      return {
        error: DendronError.createFromStatus({
          status: ERROR_STATUS.INVALID_STATE,
          message: "workspace vaults not supported",
        }),
      };
    }
    return {
      error: undefined,
    };
  }
}
