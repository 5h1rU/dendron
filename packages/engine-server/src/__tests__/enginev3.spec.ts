import {
  DEngineV2,
  DNodeUtilsV2,
  DVault,
  NotePropsV2,
  NoteUtilsV2,
  SchemaModulePropsV2,
  WorkspaceOpts,
} from "@dendronhq/common-all";
import {
  EngineTestUtilsV3,
  ENGINE_SERVER,
  NodeTestPresetsV2,
  runEngineTest,
  runJestHarnessV2,
  TestResult,
} from "@dendronhq/common-test-utils";
import { NotePresetsUtils } from "@dendronhq/common-test-utils/lib/presets/utils";
import fs from "fs-extra";
import _ from "lodash";
import { DendronEngineV2 } from "../enginev2";

const { INIT } = ENGINE_SERVER.ENGINE_MULTI_TEST_PRESET;

const setupCase1 = async () => {
  const vaults = await EngineTestUtilsV3.setupVaults({
    initVault1: async (vaultDir: string) => {
      await NotePresetsUtils.createBasic({ vaultDir, fname: "foo" });
    },
    initVault2: async (vaultDir: string) => {
      await NotePresetsUtils.createBasic({ vaultDir, fname: "bar" });
    },
  });
  const engine = DendronEngineV2.createV3({ vaults });
  return { vaults, engine };
};

const createEngine = ({ vaults }: WorkspaceOpts) => {
  return DendronEngineV2.createV3({ vaults });
};

describe("engine, notes/", () => {
  describe("write/", () => {
    test("add domain to second vault", async () => {
      await runEngineTest(
        async ({ engine, vaults }) => {
          let note = NoteUtilsV2.create({ fname: "alpha", vault: vaults[1] });
          await engine.writeNote(note);
          const root = NoteUtilsV2.getNotesByFname({
            fname: "root",
            notes: engine.notes,
            vault: vaults[1],
          })[0] as NotePropsV2;
          const newNote = NoteUtilsV2.getNotesByFname({
            fname: "alpha",
            notes: engine.notes,
            vault: vaults[1],
          })[0] as NotePropsV2;
          const results = [
            {
              actual: root.children.length,
              expected: 2,
            },
            {
              actual: newNote.parent,
              expected: root.id,
            },
          ] as TestResult[];
          await runJestHarnessV2(results, expect);
        },
        {
          createEngine,
        }
      );
    });
  });

  describe("init/", () => {
    // @ts-ignore
    let vaults: DVault[];
    let engine: DEngineV2;

    beforeEach(async () => {
      ({ vaults, engine } = await setupCase1());
    });

    test("basic/", async () => {
      await engine.init();
      // check schema correct
      expect(_.size(engine.schemas)).toEqual(3);
      expect(
        _.map(
          _.reject(
            _.values(engine.schemas),
            DNodeUtilsV2.isRoot
          ) as SchemaModulePropsV2[],
          ({ fname, vault }) => ({
            fname,
            vault,
          })
        ).sort()
      ).toEqual([
        {
          fname: "bar",
          vault: vaults[1],
        },
        {
          fname: "foo",
          vault: vaults[0],
        },
      ]);
      // check note correct
      expect(
        _.sortBy(
          _.reject(
            _.map(_.values(engine.notes), ({ fname, vault }) => ({
              fname,
              vault,
            })),
            DNodeUtilsV2.isRoot
          ),
          ["fname"]
        )
      ).toEqual([
        {
          fname: "bar",
          vault: vaults[1],
        },
        {
          fname: "bar.ch1",
          vault: vaults[1],
        },
        {
          fname: "foo",
          vault: vaults[0],
        },
        {
          fname: "foo.ch1",
          vault: vaults[0],
        },
      ]);
      const dir1 = fs.readdirSync(vaults[0].fsPath);
      const dir2 = fs.readdirSync(vaults[1].fsPath);
      expect(dir1).toMatchSnapshot("dir1");
      expect(dir2).toMatchSnapshot("dir2");
    });

    test("with stubs/", async () => {
      await INIT.WITH_STUBS.before({ vault: vaults[0] });
      await engine.init();
      await NodeTestPresetsV2.runJestHarness({
        opts: { notes: engine.notes, vault: vaults[0] },
        results: INIT.WITH_STUBS.results,
        expect,
      });
    });
  });
});
