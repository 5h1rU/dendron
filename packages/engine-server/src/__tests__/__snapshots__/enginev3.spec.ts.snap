// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine, notes/ init/ basic/: dir1 1`] = `"/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn"`;

exports[`engine, notes/ init/ basic/: dir1 2`] = `
Array [
  "foo.ch1.md",
  "foo.md",
  "foo.schema.yml",
  "root.md",
  "root.schema.yml",
]
`;

exports[`engine, notes/ init/ basic/: dir2 1`] = `"/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-AfOXhMM4076D"`;

exports[`engine, notes/ init/ basic/: dir2 2`] = `
Array [
  "bar.ch1.md",
  "bar.md",
  "bar.schema.yml",
  "root.md",
  "root.schema.yml",
]
`;

exports[`engine, notes/ init/ basic/: notes 1`] = `
Object {
  "4201b302-5acd-4bc1-8cf1-fca79c84fc6d": Object {
    "body": "
",
    "children": Array [
      "foo",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "4201b302-5acd-4bc1-8cf1-fca79c84fc6d",
    "links": Array [],
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": "1",
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
    },
  },
  "acc77a5b-2168-4e82-a297-2b65c8a5a84b": Object {
    "body": "
",
    "children": Array [
      "bar",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "acc77a5b-2168-4e82-a297-2b65c8a5a84b",
    "links": Array [],
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": "1",
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-AfOXhMM4076D",
    },
  },
  "bar": Object {
    "body": "
",
    "children": Array [
      "bar.ch1",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "bar",
    "id": "bar",
    "links": Array [],
    "parent": "acc77a5b-2168-4e82-a297-2b65c8a5a84b",
    "schema": Object {
      "moduleId": "bar",
      "schemaId": "bar",
    },
    "title": "Bar",
    "type": "note",
    "updated": "1",
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-AfOXhMM4076D",
    },
  },
  "bar.ch1": Object {
    "body": "
",
    "children": Array [],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "bar.ch1",
    "id": "bar.ch1",
    "links": Array [],
    "parent": "bar",
    "schema": Object {
      "moduleId": "bar",
      "schemaId": "ch1",
    },
    "title": "Ch1",
    "type": "note",
    "updated": "1",
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-AfOXhMM4076D",
    },
  },
  "foo": Object {
    "body": "
",
    "children": Array [
      "foo.ch1",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo",
    "id": "foo",
    "links": Array [],
    "parent": "4201b302-5acd-4bc1-8cf1-fca79c84fc6d",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "foo",
    },
    "title": "Foo",
    "type": "note",
    "updated": "1",
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
    },
  },
  "foo.ch1": Object {
    "body": "
",
    "children": Array [],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.ch1",
    "id": "foo.ch1",
    "links": Array [],
    "parent": "foo",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "ch1",
    },
    "title": "Ch1",
    "type": "note",
    "updated": "1",
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
    },
  },
}
`;

exports[`engine, notes/ init/ basic/: schemas 1`] = `
Object {
  "bar": Object {
    "fname": "bar",
    "imports": Array [],
    "root": Object {
      "children": Array [
        "ch1",
      ],
      "id": "bar",
      "parent": "root",
      "title": "bar",
    },
    "schemas": Object {
      "bar": Object {
        "body": "",
        "children": Array [
          "ch1",
        ],
        "created": 1605139102401,
        "data": Object {},
        "desc": "",
        "fname": "__empty",
        "id": "bar",
        "links": Array [],
        "parent": "root",
        "title": "bar",
        "type": "schema",
        "updated": 1605139102401,
        "vault": Object {
          "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-AfOXhMM4076D",
        },
      },
      "ch1": Object {
        "body": "",
        "children": Array [],
        "created": 1605139102401,
        "data": Object {},
        "desc": "",
        "fname": "__empty",
        "id": "ch1",
        "links": Array [],
        "parent": "bar",
        "title": "ch1",
        "type": "schema",
        "updated": 1605139102401,
        "vault": Object {
          "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-AfOXhMM4076D",
        },
      },
    },
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-AfOXhMM4076D",
    },
    "version": 1,
  },
  "foo": Object {
    "fname": "foo",
    "imports": Array [],
    "root": Object {
      "children": Array [
        "ch1",
      ],
      "id": "foo",
      "parent": "root",
      "title": "foo",
    },
    "schemas": Object {
      "ch1": Object {
        "body": "",
        "children": Array [],
        "created": 1605139102399,
        "data": Object {},
        "desc": "",
        "fname": "__empty",
        "id": "ch1",
        "links": Array [],
        "parent": "foo",
        "title": "ch1",
        "type": "schema",
        "updated": 1605139102399,
        "vault": Object {
          "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
        },
      },
      "foo": Object {
        "body": "",
        "children": Array [
          "ch1",
        ],
        "created": 1605139102398,
        "data": Object {},
        "desc": "",
        "fname": "__empty",
        "id": "foo",
        "links": Array [],
        "parent": "root",
        "title": "foo",
        "type": "schema",
        "updated": 1605139102398,
        "vault": Object {
          "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
        },
      },
    },
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
    },
    "version": 1,
  },
  "root": Object {
    "fname": "root",
    "imports": Array [],
    "root": Object {
      "children": Array [],
      "id": "root",
      "parent": "root",
      "title": "root",
    },
    "schemas": Object {
      "root": Object {
        "body": "",
        "children": Array [],
        "created": 1605139102399,
        "data": Object {},
        "desc": "",
        "fname": "__empty",
        "id": "root",
        "links": Array [],
        "parent": "root",
        "title": "root",
        "type": "schema",
        "updated": 1605139102399,
        "vault": Object {
          "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
        },
      },
    },
    "vault": Object {
      "fsPath": "/var/folders/65/c9lrx0tx6k7b_zlb4v1bs2rr0000gn/T/tmp-87151-6Wj27ifAa4yn",
    },
    "version": 1,
  },
}
`;
